### PR TITLE
Reuse generic Watch test for watchcache

### PR DESCRIPTION
**Pull request by wojtek-t**
Ref https://github.com/kubernetes/kubernetes/issues/109831

```release-note
NONE
```

/kind cleanup
/priority important-longterm
/sig api-machinery

/assign @stevekuznetsov

### CODE REVIEW

**Comment by wojtek-t** [comment]
@stevekuznetsov - friend ping for the review

File: staging/src/k8s.io/apiserver/pkg/storage/testing/utils.go [hunk 1]
```diff
@@ -210,6 +215,32 @@ func resourceVersionNotOlderThan(sentinel string) func(string) error {
...
+}
+
+func (s *StorageInjectingListErrors) GetList(ctx context.Context, key string, opts storage.ListOptions, listObj runtime.Object) error {
+	s.lock.Lock()
+	if s.Errors > 0 {
+		s.Errors--
+		s.lock.Unlock()
+		return fmt.Errorf("injected error")
+	}
+	s.lock.Unlock()
```
Comment (MadhavJivrajani) [inline staging/src/k8s.io/apiserver/pkg/storage/testing/utils.go:233]: nit: can we `defer` this `Unlock`?

File: staging/src/k8s.io/apiserver/pkg/storage/tests/cacher_test.go [hunk 2]
```diff
@@ -906,5 +905,11 @@ func testSetup(t *testing.T, opts ...setupOption) (context.Context, *cacherstora
 		server.Terminate(t)
 	}
 
+	// Since some tests depend on the fact that GetList shouldn't fail,
+	// we wait until the erro from the underlying storage is consumed.
```
Comment (MadhavJivrajani) [inline staging/src/k8s.io/apiserver/pkg/storage/tests/cacher_test.go:909]:
```suggestion
	// we wait until the error from the underlying storage is consumed.
```
nit

**Review by MadhavJivrajani — COMMENTED** [review]
just a couple of nits
thanks!

File: staging/src/k8s.io/apiserver/pkg/storage/tests/cacher_test.go (hunk 2 above)
Comment (wojtek-t) [reply staging/src/k8s.io/apiserver/pkg/storage/tests/cacher_test.go:909]: done

**Review by wojtek-t — COMMENTED** [review]
@MadhavJivrajani - PTAL

**Comment by wojtek-t** [comment]
/retest

@MadhavJivrajani - PTAL

**Comment by MadhavJivrajani** [comment]
/retest
flake: https://github.com/kubernetes/kubernetes/issues/114852

**Review by MadhavJivrajani — APPROVED** [review]
/lgtm
Thanks!

**Comment by MadhavJivrajani** [comment]
/retest
(infra issue)

**Pull request merged**